File: frontend/src/pages/Departments.js
/**
 * pages/Departments.js
 * Department management and details panel.
 * Admins: Full CRUD + View Details.
 * Employees: View Directory + View Details.
 */
import { useState, useEffect, useCallback } from 'react';
import { useAuth } from '../context/AuthContext';
import RoleRoute from '../components/common/RoleRoute';
import Modal from '../components/common/Modal';
import Spinner from '../components/common/Spinner';
import StatCard from '../components/common/StatCard';
import Badge from '../components/common/Badge';
import { useToast } from '../components/common/Toast';
import {
  BarChart, Bar, PieChart, Pie, Cell, Tooltip,
  ResponsiveContainer, XAxis, YAxis, Legend
} from 'recharts';
import {
  getDepartments, getDepartmentDetails, createDepartment,
  updateDepartment, deleteDepartment
} from '../services/api';
import {
  EditIcon, TrashIcon, PlusIcon, SaveIcon, BuildingIcon,
  UsersIcon, ClockIcon, DocumentTextIcon, ClipboardCheckIcon, CalendarIcon
} from '../components/common/Icons';

// Custom lightweight back arrow icon
const ArrowLeftIcon = ({ style = {} }) => (
  <svg style={{ width: 16, height: 16, display: 'inline-block', verticalAlign: 'middle', ...style }} fill="none" viewBox="0 0 24 24" strokeWidth={2} stroke="currentColor">
    <path strokeLinecap="round" strokeLinejoin="round" d="M10.5 19.5 3 12m0 0 7.5-7.5M3 12h18" />
  </svg>
);

function DepartmentDetails({ deptId, onBack }) {
  const toast = useToast();
  const [data, setData] = useState(null);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    (async () => {
      setLoading(true);
      try {
        const res = await getDepartmentDetails(deptId);
        setData(res.data);
      } catch (err) {
        toast.error('Failed to load department details.');
      } finally {
        setLoading(false);
      }
    })();
  }, [deptId, toast]);

  if (loading) return <Spinner />;
  if (!data) return <p style={{ textAlign: 'center', padding: 24 }}>Department not found.</p>;

  const { department, stats, employees, tasks, leaves, meetings } = data;

  // Chart structures
  const leaveData = [
    { name: 'Pending', value: leaves.pending, color: '#f59e0b' },
    { name: 'Approved', value: leaves.approved, color: '#10b981' },
    { name: 'Rejected', value: leaves.rejected, color: '#f43f5e' },
  ];

  const taskData = [
    { name: 'Pending', value: tasks.pending, color: '#f59e0b' },
    { name: 'In Progress', value: tasks.in_progress, color: '#6366f1' },
    { name: 'Completed', value: tasks.completed, color: '#10b981' }
  ];

  return (
    <div className="fade-in">
      {/* Header with Back button */}
      <div className="page-header" style={{ marginBottom: 20 }}>
        <div className="page-header-left">
          <button className="btn btn-secondary btn-sm" onClick={onBack} style={{ marginBottom: 12 }}>
            <ArrowLeftIcon style={{ marginRight: 6 }} /> Back to Departments
          </button>
          <h2>🏢 {department.name} Details</h2>
          <p className="td-muted">Department ID: DEP-{department.id} · Created: {new Date(department.created_at).toLocaleDateString()}</p>
        </div>
      </div>

      {/* Description Card */}
      {department.description && (
        <div className="card" style={{ marginBottom: 24, padding: '16px 20px' }}>
          <p style={{ margin: 0, color: 'var(--text-secondary)' }}><strong>Description:</strong> {department.description}</p>
        </div>
      )}

      {/* KPI Cards */}
      <div className="stats-grid" style={{ marginBottom: 24 }}>
        <StatCard icon={<UsersIcon />} value={stats.total_employees} label="Total Employees" color="#6366f1" />
        <StatCard icon={<ClockIcon />} value={stats.present_today} label="Present Today" color="#10b981" />
        <StatCard icon={<DocumentTextIcon />} value={stats.on_leave_today} label="On Leave" color="#f59e0b" />
        <StatCard icon={<ClipboardCheckIcon />} value={stats.pending_tasks} label="Pending Tasks" color="#ef4444" />
        <StatCard icon={<ClipboardCheckIcon />} value={stats.completed_tasks} label="Completed Tasks" color="#10b981" />
        <StatCard icon={<CalendarIcon />} value={stats.upcoming_meetings_count} label="Upcoming Meetings" color="#a78bfa" />
      </div>

      {/* Row 1: Charts Row */}
      <div className="grid-2" style={{ marginBottom: 24 }}>
        {/* Task Progress Distribution */}
        <div className="chart-card">
          <div className="chart-title">✅ Task Progress & Completion</div>
          {tasks.total === 0 ? (
            <div className="empty-state" style={{ height: 220, display: 'flex', alignItems: 'center', justifyContent: 'center' }}>
              <p className="td-muted">No task metrics available</p>
            </div>
          ) : (
            <ResponsiveContainer width="100%" height={220}>
              <BarChart data={taskData} margin={{ top: 4, right: 16, left: -10, bottom: 10 }}>
                <XAxis dataKey="name" tick={{ fill: '#94a3b8', fontSize: 11 }} />
                <YAxis tick={{ fill: '#94a3b8', fontSize: 11 }} allowDecimals={false} />
                <Tooltip />
                <Bar dataKey="value" name="Tasks" radius={[4, 4, 0, 0]}>
                  {taskData.map((entry, i) => <Cell key={i} fill={entry.color} />)}
                </Bar>
              </BarChart>
            </ResponsiveContainer>
          )}
        </div>

        {/* Leave status distribution */}
        <div className="chart-card">
          <div className="chart-title">📅 Leave Request Distribution</div>
          {leaves.pending + leaves.approved + leaves.rejected === 0 ? (
            <div className="empty-state" style={{ height: 220, display: 'flex', alignItems: 'center', justifyContent: 'center' }}>
              <p className="td-muted">No leave metrics available</p>
            </div>
          ) : (
            <ResponsiveContainer width="100%" height={220}>
              <PieChart>
                <Pie data={leaveData} dataKey="value" nameKey="name" cx="50%" cy="50%"
                  outerRadius={80} innerRadius={45} paddingAngle={3}
                >
                  {leaveData.map((entry, i) => <Cell key={i} fill={entry.color} />)}
                </Pie>
                <Tooltip />
                <Legend wrapperStyle={{ color: '#94a3b8', fontSize: 13 }} />
              </PieChart>
            </ResponsiveContainer>
          )}
        </div>
      </div>

      {/* Grid 2: Employees Section */}
      <div className="card" style={{ marginBottom: 24 }}>
        <div className="card-header">
          <div className="card-title">👥 Department Employee Directory</div>
        </div>
        {employees.length === 0 ? (
          <div className="empty-state">
            <div className="empty-state-icon"><UsersIcon style={{ width: 48, height: 48 }} /></div>
            <h3>No employees registered in this department</h3>
          </div>
        ) : (
          <div className="table-wrapper">
            <table>
              <thead>
                <tr>
                  <th>Employee ID</th>
                  <th>Name</th>
                  <th>Email</th>
                  <th>Role</th>
                  <th>Rank/Designation</th>
                  <th>Status</th>
                </tr>
              </thead>
              <tbody>
                {employees.map((e) => (
                  <tr key={e.id}>
                    <td className="td-muted">{e.employee_id}</td>
                    <td style={{ fontWeight: 600 }}>{e.name}</td>
                    <td className="td-muted">{e.email}</td>
                    <td><Badge status={e.is_line_manager ? 'Line Manager' : e.role} /></td>
                    <td className="td-muted">{e.rank || '—'}</td>
                    <td><Badge status={e.is_active ? 'Active' : 'Inactive'} /></td>
                  </tr>
                ))}
              </tbody>
            </table>
          </div>
        )}
      </div>

      {/* Meetings Section */}
      <div className="card">
        <div className="card-header">
          <div className="card-title">📅 Upcoming Department Meetings</div>
        </div>
        {meetings.length === 0 ? (
          <div className="empty-state">
            <div className="empty-state-icon"><CalendarIcon style={{ width: 48, height: 48 }} /></div>
            <h3>No upcoming meetings scheduled</h3>
          </div>
        ) : (
          <div className="table-wrapper">
            <table>
              <thead>
                <tr>
                  <th>Meeting Title</th>
                  <th>Scheduled Date/Time</th>
                  <th>Organizer</th>
                  <th>Virtual Link</th>
                </tr>
              </thead>
              <tbody>
                {meetings.map((m) => (
                  <tr key={m.id}>
                    <td style={{ fontWeight: 600 }}>{m.title}</td>
                    <td className="td-muted">{new Date(m.scheduled_at).toLocaleString()}</td>
                    <td className="td-muted">{m.creator_name || 'Admin'}</td>
                    <td>
                      {m.link ? (
                        <a href={m.link} target="_blank" rel="noreferrer" className="btn btn-secondary btn-sm" style={{ padding: '4px 8px', fontSize: 11 }}>
                          Join Meeting
                        </a>
                      ) : '—'}
                    </td>
                  </tr>
                ))}
              </tbody>
            </table>
          </div>
        )}
      </div>
    </div>
  );
}

function DepartmentsContent() {
  const { isAdmin } = useAuth();
  const toast = useToast();
  const [departments, setDepartments] = useState([]);
  const [loading, setLoading] = useState(true);
  const [selectedDeptId, setSelectedDeptId] = useState(null);

  // Forms and Modals
  const [showForm, setShowForm] = useState(false);
  const [editTarget, setEditTarget] = useState(null);
  const [showDelete, setShowDelete] = useState(null);
  const [form, setForm] = useState({ name: '', description: '' });
  const [saving, setSaving] = useState(false);

  const load = useCallback(async () => {
    setLoading(true);
    try {
      const { data } = await getDepartments();
      setDepartments(data);
    } catch {
      toast.error('Failed to load departments.');
    } finally {
      setLoading(false);
    }
  }, [toast]);

  useEffect(() => { load(); }, [load]);

  const openAdd = () => { setEditTarget(null); setForm({ name: '', description: '' }); setShowForm(true); };
  const openEdit = (d) => { setEditTarget(d); setForm({ name: d.name, description: d.description || '' }); setShowForm(true); };

  const handleSubmit = async (e) => {
    e.preventDefault();
    setSaving(true);
    try {
      if (editTarget) {
        await updateDepartment(editTarget.id, form);
        toast.success('Department updated.');
      } else {
        await createDepartment(form);
        toast.success('Department created.');
      }
      setShowForm(false);
      load();
    } catch (err) {
      toast.error(err.response?.data?.error || 'Operation failed.');
    } finally {
      setSaving(false);
    }
  };

  const handleDelete = async () => {
    try {
      await deleteDepartment(showDelete.id);
      toast.success('Department deleted.');
      setShowDelete(null);
      load();
    } catch (err) {
      toast.error(err.response?.data?.error || 'Cannot delete department.');
    }
  };

  if (loading) return <Spinner />;

  if (selectedDeptId) {
    return <DepartmentDetails deptId={selectedDeptId} onBack={() => setSelectedDeptId(null)} />;
  }

  return (
    <div className="fade-in">
      <div className="page-header">
        <div className="page-header-left">
          <h2>Departments</h2>
          <p>{departments.length} department{departments.length !== 1 ? 's' : ''} configured</p>
        </div>
        {isAdmin && (
          <div className="page-header-actions">
            <button className="btn btn-primary" onClick={openAdd}><PlusIcon style={{ marginRight: 6 }} /> Add Department</button>
          </div>
        )}
      </div>

      <div className="card">
        <div className="table-wrapper">
          <table>
            <thead>
              <tr>
                <th>#</th>
                <th>Name</th>
                <th>Description</th>
                <th>Created</th>
                <th>Actions</th>
              </tr>
            </thead>
            <tbody>
              {departments.length === 0 && (
                <tr>
                  <td colSpan={5}>
                    <div className="empty-state">
                      <div className="empty-state-icon"><BuildingIcon style={{ width: 48, height: 48 }} /></div>
                      <h3>No departments yet</h3>
                    </div>
                  </td>
                </tr>
              )}
              {departments.map((d, i) => (
                <tr key={d.id}>
                  <td className="td-muted">{i + 1}</td>
                  <td style={{ fontWeight: 600 }}>{d.name}</td>
                  <td className="td-muted">{d.description || '—'}</td>
                  <td className="td-muted">{new Date(d.created_at).toLocaleDateString()}</td>
                  <td>
                    <div className="table-actions">
                      <button className="btn btn-secondary btn-sm" onClick={() => setSelectedDeptId(d.id)}>
                        View Details
                      </button>
                      {isAdmin && (
                        <>
                          <button className="btn btn-secondary btn-sm" onClick={() => openEdit(d)}>
                            <EditIcon style={{ marginRight: 6 }} /> Edit
                          </button>
                          <button className="btn btn-danger btn-sm" onClick={() => setShowDelete(d)}>
                            <TrashIcon style={{ marginRight: 6 }} /> Delete
                          </button>
                        </>
                      )}
                    </div>
                  </td>
                </tr>
              ))}
            </tbody>
          </table>
        </div>
      </div>

      {/* Add/Edit Modal */}
      <Modal isOpen={showForm} onClose={() => setShowForm(false)} title={editTarget ? `Edit: ${editTarget.name}` : 'New Department'} size="sm">
        <form onSubmit={handleSubmit}>
          <div className="form-group">
            <label className="form-label">Department Name <span className="form-required">*</span></label>
            <input className="form-control" value={form.name} onChange={(e) => setForm((f) => ({ ...f, name: e.target.value }))} required placeholder="e.g. Engineering" />
          </div>
          <div className="form-group">
            <label className="form-label">Description</label>
            <textarea className="form-control" value={form.description} onChange={(e) => setForm((f) => ({ ...f, description: e.target.value }))} rows={3} />
          </div>
          <div className="form-actions">
            <button type="button" className="btn btn-secondary" onClick={() => setShowForm(false)} disabled={saving}>Cancel</button>
            <button type="submit" className="btn btn-primary" disabled={saving}>
              {saving ? 'Saving…' : editTarget ? <><SaveIcon style={{ marginRight: 6 }} /> Update</> : <><PlusIcon style={{ marginRight: 6 }} /> Create</>}
            </button>
          </div>
        </form>
      </Modal>

      {/* Delete Confirm */}
      <Modal isOpen={!!showDelete} onClose={() => setShowDelete(null)} title="Delete Department" size="sm">
        <p style={{ color: 'var(--text-secondary)' }}>
          Delete <strong>{showDelete?.name}</strong>? This cannot be undone.
          Departments with active employees cannot be deleted.
        </p>
        <div className="form-actions">
          <button className="btn btn-secondary" onClick={() => setShowDelete(null)}>Cancel</button>
          <button className="btn btn-danger" onClick={handleDelete}><TrashIcon style={{ marginRight: 6 }} /> Delete</button>
        </div>
      </Modal>
    </div>
  );
}

export default function Departments() {
  return <RoleRoute roles={['Admin', 'Employee']}><DepartmentsContent /></RoleRoute>;
}

File: backend/routes/departments.py
"""
backend/routes/departments.py
------------------------------
Department management endpoints.

Endpoints:
  GET  /api/departments        - List all departments (authenticated)
  POST /api/departments        - Admin: Create a department
  PUT  /api/departments/<id>   - Admin: Update a department
  DELETE /api/departments/<id> - Admin: Delete a department
"""

from flask import Blueprint, request, jsonify
from models import Department
from extensions import db
from utils.decorators import jwt_required, admin_required

departments_bp = Blueprint("departments", __name__)

# Default departments to seed on first run (handled in init_db.py)
DEFAULT_DEPARTMENTS = [
    "Research", "Tech", "GIS", "Data Scientist",
    "Broker", "Execution", "Account", "Management",
]


# ------------------------------------------------------------------
# GET /api/departments
# ------------------------------------------------------------------
@departments_bp.route("/", methods=["GET"])
@jwt_required
def list_departments(current_user_id, current_user_role):
    """Return all departments."""
    departments = Department.query.order_by(Department.name).all()
    return jsonify([d.to_dict() for d in departments]), 200


# ------------------------------------------------------------------
# POST /api/departments
# ------------------------------------------------------------------
@departments_bp.route("/", methods=["POST"])
@admin_required
def create_department(current_user_id, current_user_role):
    """Admin-only: Create a new department."""
    data = request.get_json(silent=True)
    if not data or not data.get("name"):
        return jsonify({"error": "'name' is required."}), 422

    name = data["name"].strip()
    if Department.query.filter_by(name=name).first():
        return jsonify({"error": f"Department '{name}' already exists."}), 409

    dept = Department(name=name, description=data.get("description"))
    db.session.add(dept)
    db.session.commit()
    return jsonify(dept.to_dict()), 201


# ------------------------------------------------------------------
# PUT /api/departments/<id>
# ------------------------------------------------------------------
@departments_bp.route("/<int:dept_id>", methods=["PUT"])
@admin_required
def update_department(dept_id, current_user_id, current_user_role):
    """Admin-only: Update department name or description."""
    dept = Department.query.get_or_404(dept_id)
    data = request.get_json(silent=True) or {}

    if "name" in data:
        dept.name = data["name"].strip()
    if "description" in data:
        dept.description = data["description"]

    db.session.commit()
    return jsonify(dept.to_dict()), 200


# ------------------------------------------------------------------
# DELETE /api/departments/<id>
# ------------------------------------------------------------------
@departments_bp.route("/<int:dept_id>", methods=["DELETE"])
@admin_required
def delete_department(dept_id, current_user_id, current_user_role):
    """Admin-only: Delete a department (only if no employees assigned)."""
    dept = Department.query.get_or_404(dept_id)

    if dept.employees.count() > 0:
        return jsonify({
            "error": "Cannot delete department with assigned employees. Reassign them first."
        }), 409

    db.session.delete(dept)
    db.session.commit()
    return jsonify({"message": f"Department '{dept.name}' deleted."}), 200


# ------------------------------------------------------------------
# GET /api/departments/<id>
# ------------------------------------------------------------------
@departments_bp.route("/<int:dept_id>", methods=["GET"])
@jwt_required
def get_department_details(dept_id, current_user_id, current_user_role):
    """Return department aggregated details and statistics."""
    from datetime import date, datetime
    from models import User, Attendance, Leave, Task, Meeting

    dept = Department.query.get_or_404(dept_id)

    # 1. Employees
    employees = dept.employees.all()
    emp_ids = [e.id for e in employees]

    # 2. Today's attendance
    today = date.today()
    present_today = 0
    if emp_ids:
        present_today = Attendance.query.filter(
            Attendance.employee_id.in_(emp_ids),
            Attendance.date == today,
            Attendance.check_in != None
        ).count()

    # 3. Employees on leave today
    on_leave_today = 0
    if emp_ids:
        on_leave_today = Leave.query.filter(
            Leave.employee_id.in_(emp_ids),
            Leave.status == "Approved",
            Leave.start_date <= today,
            Leave.end_date >= today
        ).count()

    # 4. Tasks Stats
    total_tasks = 0
    pending_tasks = 0
    in_progress_tasks = 0
    completed_tasks = 0
    if emp_ids:
        total_tasks = Task.query.filter(Task.employee_id.in_(emp_ids)).count()
        pending_tasks = Task.query.filter(Task.employee_id.in_(emp_ids), Task.status == "Pending").count()
        in_progress_tasks = Task.query.filter(Task.employee_id.in_(emp_ids), Task.status == "In Progress").count()
        completed_tasks = Task.query.filter(Task.employee_id.in_(emp_ids), Task.status == "Completed").count()

    # 5. Leaves Stats
    pending_leaves = 0
    approved_leaves = 0
    rejected_leaves = 0
    if emp_ids:
        pending_leaves = Leave.query.filter(Leave.employee_id.in_(emp_ids), Leave.status == "Pending").count()
        approved_leaves = Leave.query.filter(Leave.employee_id.in_(emp_ids), Leave.status == "Approved").count()
        rejected_leaves = Leave.query.filter(Leave.employee_id.in_(emp_ids), Leave.status == "Rejected").count()

    # 6. Meetings
    now = datetime.utcnow()
    upcoming_meetings_query = Meeting.query.filter(
        Meeting.department_id == dept.id,
        Meeting.scheduled_at >= now
    )
    upcoming_meetings_count = upcoming_meetings_query.count()
    upcoming_meetings = upcoming_meetings_query.order_by(Meeting.scheduled_at.asc()).all()

    # Serialize employees
    serialized_employees = []
    for e in employees:
        serialized_employees.append({
            "id": e.id,
            "employee_id": e.employee_id,
            "name": e.name,
            "email": e.email,
            "role": e.role,
            "rank": e.rank,
            "is_active": e.is_active
        })

    # Serialize meetings
    serialized_meetings = [m.to_dict() for m in upcoming_meetings]

    return jsonify({
        "department": dept.to_dict(),
        "stats": {
            "total_employees": len(employees),
            "present_today": present_today,
            "on_leave_today": on_leave_today,
            "pending_tasks": pending_tasks,
            "completed_tasks": completed_tasks,
            "upcoming_meetings_count": upcoming_meetings_count
        },
        "employees": serialized_employees,
        "tasks": {
            "total": total_tasks,
            "pending": pending_tasks,
            "in_progress": in_progress_tasks,
            "completed": completed_tasks
        },
        "leaves": {
            "pending": pending_leaves,
            "approved": approved_leaves,
            "rejected": rejected_leaves
        },
        "meetings": serialized_meetings
    }), 200

